server/controllers/medical/patients/index.js

```diff
@@ -423,6 +427,289 @@ function searchByName(req, res, next) {
+ * minor misspellings.
+ *
+ * This function supports these query parameters:
+ *    - search_name      string [required]
+ *    - sex              string [optional] ('F' / 'M')
+ *    - dob              string [optional] (see note below)
+ *    - dob_unknown_date string [optional] ('true'/'false')
+ *
+ * Note that dob can be an SQL date format string, a four-digit string
+ * for the year (if dob_unknown_date is true), or a 4-digit number for
+ * the year.
+ *
+ * The goal is to make the approximate name search work as robustly
+ * as possible.
+ *
+ * The first step is to find potential patient name matches:
+ *   - The search_name and potential matching patient names are split
+ *     into name parts (eg, first  name, last name, etc) and
+ *     alphabetized.
+ *   - Then the names are compared in the alphabetical order and the
+ *     'distance' (0-1) between them is determined using the
+ *     jaro-winkler function.  This function gives exact matches
+ *     if the names are the same (ignoring capitalization) and is
+ *     tolerant of the order of the name parts.  Note that a
+ *     variation of this approach is used if there are different
+ *     numbers of name parts for the search and patient names.
+ *   - The total score for the name is based sum of the distances
+ *     between each pair of name parts divided by the number of parts
+ *     in the search name.   The result is always between 0 and 1.
+ *   - Matches that are off single or double misspellings generally
+ *     score very close to 1.
+ *   - The function findMatchingPatients() does the name part
+ *     comparisons to construct a list of potential name matches,
+ *     along with the score of each match.  Matches with scores above
+ *     a cut off value are kept as a potential match is 'matchCutoff'
+ *     (which is defined above in findMatchingPatients()).
+ *
+ * Once the potential name matches are determined, then the optional
+ * gender and date of birth fields come into play.
+ *   - If the 'sex' field is present, it is matched and the score is
+ *     augmented by 50% (see sexWeight) below.
+ *   - If the 'dob' field is present, it is also matched and the score
+ *     is augmented by a maximum of 30% (see dobWeight below).  DOB
+ *     matches that are exact count the maximum. Matches that are
+ *     approximate are discounted the further off they are.  Currently
+ *     dob the match needs to be within a few year for a reasonable
+ *     increment.  There is extra logic below that deals with year
+ *     year matches vs exact date matches.
+ *   - In any case, the total score for each potential match is
+ *     scaled to be between 0 and 1.
+ *   - Specifying the sex and DOB can help to improve the ranking
+ *     of potential matches.
+ *
+ * @returns array of potential matching patient objects.
+ *          Each row has a additional 'matchScore' value that
+ *          indicates the quality of the match (0-1).
+ */
```

Comment:
I think we need to include length as a really (minor) criteria.  For example, I can search for "jean" and get many 100% matches (jean jacques niles would be a match).  This doesn't totally make sense to me - we aren't 100% confident that is a match.

If we could add a weighting to the name so that the more information (3 name parts > 2 name parts > 1 name) would increase the score, that would be helpful.  Do you think that is possible?

Reply:
I've added this functionality, updated the tests, and added a new test for this feature.